core/async/src/test_loop/futures.rs

```diff
@@ -1,223 +1,107 @@
+///      future, which will resolve whenever the TestLoop handles the message.
+///      To support this, use MessageWithCallback<Request, Response> as the
+///      event type, and in the handler, call (event.responder)(result)
+///      (possibly asynchronously) to complete the future.
+///
+///      This is needed to support the AsyncSender interface, which is required
+///      by some components as they expect a response to each message. The way
+///      this is implemented is by implementing a conversion from
+///      DelaySender<MessageWithCallback<Request, Response>> to
+///      AsyncSender<Request, Response>.
 
-/// A DelaySender<Arc<TestLoopTask>> is a FutureSpawner that can be used to
+/// A DelaySender is a FutureSpawner that can be used to
 /// spawn futures into the test loop. We give it a convenient alias.
-pub type TestLoopFutureSpawner = DelaySender<Arc<TestLoopTask>>;
+pub type TestLoopFututeSpawner = DelaySender;
 
-impl FutureSpawner for TestLoopFutureSpawner {
+impl FutureSpawner for TestLoopFututeSpawner {
     fn spawn_boxed(&self, description: &str, f: BoxFuture<'static, ()>) {
-        let task = Arc::new(TestLoopTask {
-            future: Mutex::new(Some(f)),
-            sender: self.clone(),
-            description: description.to_string(),
-        });
-        self.send(task);
-    }
-}
-
-/// Represents an action that was scheduled to run later, by using
-/// `DelayedActionRunner::run_later`.
-pub struct TestLoopDelayedActionEvent<T> {
-    name: String,
-    action: Box<dyn FnOnce(&mut T, &mut dyn DelayedActionRunner<T>) + Send + 'static>,
-}
-
-impl<T> Debug for TestLoopDelayedActionEvent<T> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_tuple("DelayedAction").field(&self.name).finish()
-    }
-}
-
-/// An event handler that handles only `TestLoopDelayedActionEvent`s, by
-/// running the action encapsulated in the event.
-pub fn drive_delayed_action_runners<T>(
-    sender: DelaySender<TestLoopDelayedActionEvent<T>>,
-    shutting_down: Arc<AtomicBool>,
-) -> LoopEventHandler<T, TestLoopDelayedActionEvent<T>> {
-    LoopEventHandler::new_simple(move |event: TestLoopDelayedActionEvent<T>, data: &mut T| {
-        let mut runner = TestLoopDelayedActionRunner {
-            sender: sender.clone(),
-            shutting_down: shutting_down.clone(),
+        let task = Arc::new(FutureTask { future: Mutex::new(Some(f)), sender: self.clone() });
+        let callback = move |_: &mut TestLoopData| {
+            drive_futures(&task);
         };
-        (event.action)(data, &mut runner);
-    })
-}
-
-/// `DelayedActionRunner` that schedules the action to be run later by the
-/// TestLoop event loop.
-pub struct TestLoopDelayedActionRunner<T> {
-    pub(crate) sender: DelaySender<TestLoopDelayedActionEvent<T>>,
-    pub(crate) shutting_down: Arc<AtomicBool>,
-}
-
-impl<T> DelayedActionRunner<T> for TestLoopDelayedActionRunner<T> {
```

Comment:
DelayedActionRunner is implemented by the TestLoopSender now. It is specific to actors.